### ISSUE TITLE
SteamVR passthrough support

### ISSUE THREAD

**Issue by Rectus** [mbucchia/WMR-Passthrough#1]
SteamVR exposes passthrough camera data for most native SteamVR headsets through its OpenVR IVRTrackedCamera API, but does not support OpenXR passthrough for some reason. 

I've been thinking about doing something like this API layer for the SteamVR runtime, but haven't gotten to it. I was wondering if you had any thoughts about adding SteamVR support for this. 

Otherwise I could try my hand on forking it and adding support for it if/when I'm able.
I've already implemented it once as an Unreal engine plugin:
https://github.com/Rectus/UE4SteamVRPassthrough

**Comment by mbucchia** [comment]
Hey!

Thanks for your message! This OpenXR API layer isn't finished yet, I can't find time to finish it (support the Reverb G2, which requires to modify the XRmonitors camera hack, and solve some of the latency issues).

I have no plans to support SteamVR because I'm fully focused on OpenXR.

**Comment by Rectus** [comment]
> I have no plans to support SteamVR because I'm fully focused on OpenXR.

Ah sorry, I meant adding support for the SteamVR OpenXR runtime.

**Comment by mbucchia** [comment]
I'm not sure I see the use case. I always encourage people to use the native OpenXR runtime. For the applications of passthrough, I think supporting Unity and other AR-capable engines is enough, and I'm not sure why one would run these with SteamVR.

**Comment by Rectus** [comment]
I mean for headsets that are native to SteamVR. I think all SteamVR native HTC and Valve headsets have passthrough support, but can't be run through any other runtime than SteamVR officially.

**Comment by mbucchia** [comment]
Oh, I think you misunderstood what my project does: it captures the camera exclusively for WMR devices. The APIs (or hacks) to use cameras don't have any common standard. So this needs to be implemented per-headset, and I only plan to support WMR for this project, and OpenXR API layer only.

**Comment by Rectus** [comment]
Then I think I do understand it correctly. My idea was to add support for SteamVR devices as well as WMR devices to the OpenXR API layer, since neither runtime has OpenXR passthrough support natively. 

This should be relatively straightforward to implement, since the old SteamVR API has a fairly low-level camera interface with support for everything that passes camera data to it from the driver, and it can be seamlessly initialized to run in the background with OpenXR applications.

Thanks for responses! I'll experiment with setting it up myself if I get the time.

**Comment by mbucchia** [comment]
OK so sorry I understand now, I got thrown off by the mention of the SteamVR classes and SteamVR OpenXR runtime! I thought you wanted to write a SteamVR driver. 

But it makes sense to take the API layer and plumb other camera drivers into it.

BTW, this is one of my "1st generation" API layer and I now have a much better template here: https://github.com/mbucchia/OpenXR-Layer-Template

That could be a better starting point. Then as you've probably already found from the code, it's only a matter of inserting a new layer in `xrEndFrame()` (which is drawn using the texture from the camera), with of course the difficulty to support every single graphics API (and I only supported D3D11 and 12).
